File: src/java/backtracking/B_2580.java
package java.backtracking;

import java.util.*;
import java.io.*;

public class B_2580 {
	static int[][] board = new int[9][9];
	static StringBuilder sb = new StringBuilder();

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		for(int i=0;i<9;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<9;j++) 
				board[i][j] = Integer.parseInt(st.nextToken());
		}

		DFS(0);
	}

	static void DFS(int depth) {
		if(depth == 9*9) {
			for(int i=0;i<9;i++) {
				for(int j=0;j<9;j++)
					sb.append(board[i][j]).append(' ');
				sb.append('\n');
			}
			System.out.println(sb);
			System.exit(0);
		}

		int r = depth / 9;
		int c = depth % 9;
		if(board[r][c] != 0)
			DFS(depth + 1);
		else {
			for(int i=1;i<10;i++) {
				if(check(r, c, i)) {
					board[r][c] = i;
					DFS(depth + 1);
				}
			}
			board[r][c] = 0;
		}
	}

	static boolean check(int x, int y, int value) {
		for(int i=0;i<9;i++) {
			if(board[i][y] == value)
				return false;
		}

		for(int i=0;i<9;i++) {
			if(board[x][i] == value)
				return false;
		}

		int sq_x = (x / 3) * 3;
		int sq_y = (y / 3) * 3;

		for(int i=sq_x;i<sq_x+3;i++) {
			for(int j=sq_y;j<sq_y+3;j++)
				if(board[i][j] == value)
					return false;
		}
		return true;
	}

}

File: src/kotlin/backtracking/B_2580.kt
import kotlin.system.exitProcess

val arr = Array(9) { IntArray(9) }
val zero = mutableListOf<Int>()
var n = 0

fun main() = with(System.`in`.bufferedReader()) {
    repeat(9) { i ->
        val input = readLine().split(" ")
        repeat(9) { j ->
            arr[i][j] = input[j].toInt()
            if (arr[i][j] == 0) {
                zero.add(i * 9 + j)
            }
        }
    }
    n = zero.size
    dfs(0)
}

fun dfs(depth: Int) {
    if (depth == n) {
        val sb = StringBuilder()
        repeat(9) {
            sb.append(arr[it].joinToString(" ")).append('\n')
        }
        print(sb)
        exitProcess(0)
    }

    val x = zero[depth] / 9
    val y = zero[depth] % 9
    for (i in 1..9) {
        if (!check(x, y, i)) {
            continue
        }
        arr[x][y] = i
        dfs(depth + 1)
        arr[x][y] = 0
    }
}

fun check(x: Int, y: Int, num: Int): Boolean {
    val startX = if (x in 0..2) 0 else if (x in 3..5) 3 else 6
    val startY = if (y in 0..2) 0 else if (y in 3..5) 3 else 6

    for (i in startX..startX + 2) {
        for (j in startY..startY + 2) {
            if (x == i && y == j) continue
            if (arr[i][j] == num) return false
        }
    }

    for (i in 0..8) {
        if (y == i) continue
        if (arr[x][i] == num) return false
    }

    for (i in 0..8) {
        if (x == i) continue
        if (arr[i][y] == num) return false
    }
    return true
}
